riscv/src/runtime.rs

```diff
@@ -366,9 +498,79 @@ impl TryFrom<&[&str]> for Runtime {
             }
             match *name {
                 "poseidon_gl" => runtime = runtime.with_poseidon(),
+                "arith" => runtime = runtime.with_arith(),
                 _ => return Err(format!("Invalid co-processor specified: {name}")),
             }
         }
         Ok(runtime)
     }
 }
+
+/// Helper function for register names used in instruction params
+fn reg(mut idx: u8) -> String {
+    // s* callee saved registers
+    static SAVED_REGS: [&str; 12] = [
+        "x8", "x9", "x18", "x19", "x20", "x21", "x22", "x23", "x24", "x25", "x26", "x27",
+    ];
+
+    // first, use syscall_registers
+    if idx < SYSCALL_REGISTERS.len() as u8 {
+        return SYSCALL_REGISTERS[idx as usize].to_string();
+    }
+    idx -= SYSCALL_REGISTERS.len() as u8;
+    // second, use s* registers
+    if idx < SAVED_REGS.len() as u8 {
+        return SAVED_REGS[idx as usize].to_string();
+    }
+    idx -= SAVED_REGS.len() as u8;
+    idx += 1;
+    // lastly, use extra submachine registers
+    format!("{EXTRA_REG_PREFIX}{idx}")
+}
+
+/// Helper function to generate params (i.e., "A, B -> C, D") for instruction declarations using registers
+fn instr_register_params(start_idx: u8, inputs: u8, outputs: u8) -> String {
+    format!(
+        "{} -> {}",
+        (start_idx..start_idx + inputs).map(reg).join(", "),
+        (start_idx..start_idx + outputs)
+            .map(|i| format!("{}'", reg(i)))
+            .join(", "),
+    )
+}
+
+/// Load gl field element from addr+offset into register
+fn load_gl_fe(addr: &str, offset: u32, reg: &str) -> [String; 3] {
+    let lo = offset;
+    let hi = offset + 4;
+    [
+        format!("{reg}, tmp2 <== mload({lo} + {addr});"),
+        format!("tmp1, tmp2 <== mload({hi} + {addr});"),
+        format!("{reg} <=X= {reg} + tmp1 * 2**32;"),
+    ]
+}
+
+/// Store gl field element from register into addr+offset
+fn store_gl_fe(addr: &str, offset: u32, reg: &str) -> [String; 3] {
+    let lo = offset;
+    let hi = offset + 4;
+    [
+        format!("tmp1, tmp2 <== split_gl({reg});"),
+        format!("mstore {lo} + {addr}, tmp1;"),
+        format!("mstore {hi} + {addr}, tmp2;"),
+    ]
+}
+
+/// Load word from addr+offset into register
+fn load_word(addr: &str, offset: u32, reg: &str) -> [String; 1] {
+    [format!("{reg}, tmp2 <== mload({offset} + {addr});")]
+}
+
+/// Store word from register into addr+offset
+fn store_word(addr: &str, offset: u32, reg: &str) -> [String; 2] {
+    [
+        // riscv-executor seems to need the split_gl to ensure the register has a 32-bit value
```

Comment:
it's not really the riscv-executor, but rather the invariant we try to keep in the RISCV machine that the RISCV registers (x0-x31) always have a 32-bit value in them even though technically they're FEs for us